module/zfs/spa.c

```diff
@@ -6819,10 +6819,12 @@ spa_vdev_attach(spa_t *spa, uint64_t guid, nvlist_t *nvroot, int replacing,
 		return (spa_vdev_exit(spa, newrootvd, txg, error));
 
 	/*
-	 * Spares can't replace logs
+	 * log, dedup and special vdevs should not be replaced by spares.
 	 */
-	if (oldvd->vdev_top->vdev_islog && newvd->vdev_isspare)
+	if ((oldvd->vdev_top->vdev_alloc_bias != VDEV_BIAS_NONE ||
+	    oldvd->vdev_top->vdev_islog) && newvd->vdev_isspare) {
```

Comment:
`oldvd->vdev_top->vdev_alloc_bias != VDEV_BIAS_NONE` should make it unnecessary to check `oldvd->vdev_top->vdev_islog`.

Reply:
Yes, I saw one updated on another, but IMHO it is updated in pretty weird place -- vdev_metaslab_group_create(), so I proposed to be safe rather than sorry.  It is not a performance critical path.

Reply:
The static analyzers are not smart enough to complain that this is unnecessary, so it is fine to leave this as-is.

That said, this feels like a future code cleanup opportunity.